crawl-ref/source/util/species-gen.py

```diff
@@ -0,0 +1,682 @@
+#!/usr/bin/env python3
+
+import argparse
+import os
+import sys
+import traceback
+import collections
+
+import yaml  # pip install pyyaml
+
+
+class Species(dict):
+    """Parser for YAML definition files.
+
+    If any YAML content is invalid, the relevant parser function below should
+    raise ValueError.
+    """
+    @staticmethod
+    def from_yaml(s):
+        species = Species()
+        species['tag_major_version_opener'] = '' if 'TAG_MAJOR_VERSION' not in s else "#if TAG_MAJOR_VERSION == %s" % s['TAG_MAJOR_VERSION']
+        species['tag_major_version_closer'] = '' if 'TAG_MAJOR_VERSION' not in s else "#endif"
+        species['enum'] = s.get('enum', "SP_%s" % s['name'].replace(' ', '_').upper())
+        species['name'] = s['name']
+        species['short_name'] = s.get('short_name', s['name'][:2])
+        species['adjective'] = quote(s['adjective']) if 'adjective' in s else 'nullptr'
+        species['genus'] = quote(s['genus']) if 'genus' in s else 'nullptr'
+        species['species_flags'] = species_flags(s.get('species_flags', []))
+        species['xp'] = s['aptitudes']['xp']
+        species['hp'] = s['aptitudes']['hp']
+        species['mp'] = s['aptitudes']['mp_mod']
+        species['mr'] = s['aptitudes']['mr']
+        species['aptitudes'] = aptitudes(s['aptitudes'])
+        species['monster_name'] = s.get('monster', "MONS_%s" % s['name'].replace(' ', '_').upper())
```

Comment:
Perhaps `s.replace(' ', '_').upper()` should be extracted to an `enumify()` function, which takes a prefix (`MONS_` in this example).

Reply:
I'm still not totally sold on the motivation for not just using the enum names here.  From an earlier discussion, the idea is to make things as easy as possible, which is nice and all. But from a practical perspective, no matter how easy, there won't be *that* many species proposals, and the people who have to maintain this and keep track of the mapping from enum names to data keywords is still us. So all this is to say that I'm still on the fence about introducing a whole new set of keywords for this format which are not the same as the enum names.

But if this is to be merged with this, the mapping needs to be absolutely as transparent as possible, and so abstracting this into a function is a good suggestion.

Reply:
I added an `enumify` function as suggested

Reply:
Yeah there's an argument to be made against massaging input into an enum; if so, the function should at least verify that the format is correct.

Reply:
After sleeping on it, I agree. Enum values are used elsewhere in the code, people adding or modifying a species will want to be able to find the enum name of a species somewhere other than an auto-generated file. I will change the enum value to be a required field and validate it matches `SP_[A-Z]+`